core/service.go

```diff
@@ -2,26 +2,542 @@ package core
+			srcRef, err := solveRef(ctx, gw, mnt.Source)
+			if err != nil {
+				return nil, fmt.Errorf("mount %s: %w", mnt.Target, err)
+			}
+
+			mount.Ref = srcRef
+		}
+
+		if mnt.SourcePath != "" {
+			mount.Selector = mnt.SourcePath
+		}
+
+		if mnt.CacheID != "" {
+			mount.MountType = pb.MountType_CACHE
+			mount.CacheOpt = &pb.CacheOpt{
+				ID: mnt.CacheID,
+			}
+
+			switch mnt.CacheSharingMode {
+			case "shared":
+				mount.CacheOpt.Sharing = pb.CacheSharingOpt_SHARED
+			case "private":
+				mount.CacheOpt.Sharing = pb.CacheSharingOpt_PRIVATE
+			case "locked":
+				mount.CacheOpt.Sharing = pb.CacheSharingOpt_LOCKED
+			default:
+				return nil, errors.Errorf("invalid cache mount sharing mode %q", mnt.CacheSharingMode)
+			}
+		}
+
+		if mnt.Tmpfs {
+			mount.MountType = pb.MountType_TMPFS
+		}
+
+		mounts = append(mounts, mount)
+	}
+
+	if opts.ExperimentalPrivilegedNesting {
+		env = append(env, "_DAGGER_ENABLE_NESTING=")
+	}
+
+	if opts.RedirectStdout != "" {
+		env = append(env, "_DAGGER_REDIRECT_STDOUT="+opts.RedirectStdout)
+	}
+
+	if opts.RedirectStderr != "" {
+		env = append(env, "_DAGGER_REDIRECT_STDERR="+opts.RedirectStderr)
+	}
+
+	for _, alias := range ctr.HostAliases {
+		env = append(env, "_DAGGER_HOSTNAME_ALIAS_"+alias.Alias+"="+alias.Target)
+	}
+
+	var stdin io.ReadCloser
+	if opts.Stdin != "" {
+		stdin = io.NopCloser(strings.NewReader(opts.Stdin))
+	}
+
+	var securityMode pb.SecurityMode
+	if opts.InsecureRootCapabilities {
+		securityMode = pb.SecurityMode_INSECURE
+	}
+
+	rec := progrock.RecorderFromContext(ctx)
+
+	dig, err := svc.Digest()
+	if err != nil {
+		return nil, err
+	}
+
+	host, err := svc.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	vtx := rec.Vertex(dig, "start "+strings.Join(args, " "))
+
+	gc, err := gw.NewContainer(ctx, bkgw.NewContainerRequest{
+		Mounts:   mounts,
+		Hostname: host,
+		Platform: &pbPlatform,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	proc, err := gc.Start(ctx, bkgw.StartRequest{
+		Args:         args,
+		Env:          env,
+		User:         cfg.User,
+		Cwd:          cfg.WorkingDir,
+		Tty:          false,
+		Stdin:        stdin,
+		Stdout:       nopCloser{vtx.Stdout()},
+		Stderr:       nopCloser{vtx.Stderr()},
+		SecurityMode: securityMode,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		// Keep dependent services running so long as the service is running.
+		_ = proc.Wait()
+		detach()
+	}()
+
+	return &RunningService{
+		Service:   svc,
+		Hostname:  host,
+		Container: gc,
+		Process:   proc,
+	}, nil
+}
+
+type RunningService struct {
+	Service   *Service
+	Hostname  string
+	Container bkgw.Container
+	Process   bkgw.ContainerProcess
+}
+
+type Services struct {
+	progSock *Socket
+	starting map[string]*sync.WaitGroup
+	running  map[string]*RunningService
+	bindings map[string]int
+	l        sync.Mutex
+}
+
+// AllServices is a pesky global variable storing the state of all running
+// services.
+var AllServices *Services
```

Comment:
Yeah definitely. This was a bit of a hacky step because injecting it everywhere seemed like a huge pain; basically every object needs it. I'll just make it private for now.